controllers/state_of_the_world.go

```diff
@@ -181,6 +191,11 @@ func (r *TopologyFileReconciler) Reconcile(ctx context.Context, _ []controller.R
 	if len(existingTopologyConfigMaps) == 0 {
 		_, err := r.Client.Resource(controller.ConfigMapsResource).Namespace(cm.Namespace).Create(ctx, unstructuredCM, metav1.CreateOptions{})
 		if err != nil {
+			if errors.IsAlreadyExists(err) {
+				// This error can happen when the operator is starting, and the create event for the topology has not being processed.
```

Comment:
Hmmm... We may expect more of these, and not only with this particular resource.

This is because, as a pattern, we check the presence of the resource in the topology, and not with the API server. But more importantly, because, even though it's "state of the world" reconciliation and therefore the controller already knows about all resources that exist (because it lists them all), the library watching for changes to the cache writes to the subscription channel [resource by resource](https://github.com/Kuadrant/policy-machinery/blob/585f10911abdbb919512b3317c9db073eccf99a3/controller/cache.go#L100).

I think we should consider fixing this in the Policy Machinery instead, because now it may be defeating the purpose of "state of the world". We may be able to make the call to [`cache.Replace`](https://github.com/Kuadrant/policy-machinery/blob/585f10911abdbb919512b3317c9db073eccf99a3/controller/controller.go#L220C2-L220C24) truly atomic – assuming it won't break with the granularity of multiple event types. I have a couple of ideas.

Reply:
I believe https://github.com/Kuadrant/policy-machinery/pull/35 may have fixed this issue.

Reply:
This still happens with the same error:

```
{"level":"error","ts":"2024-10-02T08:56:41+01:00","logger":"kuadrant-operator","msg":"reconciliation error","error":"configmaps \"topology\" already exists","stacktrace":"github.com/kuadrant/policy-machinery/controller.(*Controller).propagate\n\t/Users/chfan/dev/kuadrant-operator/vendor/github.com/kuadrant/policy-machinery/controller/controller.go:262\ngithub.com/kuadrant/policy-machinery/controller.(*Controller).subscribe.func2\n\t/Users/chfan/dev/kuadrant-operator/vendor/github.com/kuadrant/policy-machinery/controller/controller.go:312"}
{"level":"info","ts":"2024-10-02T08:56:41+01:00","logger":"kuadrant-operator.controller-runtime.metrics","msg":"Starting metrics server"}
{"level":"info","ts":"2024-10-02T08:56:41+01:00","logger":"kuadrant-operator","msg":"starting server","name":"health probe","addr":"[::]:8081"}
```

What looks to happens from debugging is that this code is ran on boot even when there has been no events and when the topology has been loaded with the cluster data yet 🤔 

My guess is that since this is a task with no event matchers, the initial creation of the empty cache store is an event that causes these task to run 🤔

Reply:
> the initial creation of the empty cache store is an event that causes these task to run

Oh! This makes total sense and explains why it's triggering before the controller finishes starting. By simply creating the cache we may be generating events already. Nice catch, @KevFan!

Reply:
I'm guessing that somewhere [here](https://github.com/Kuadrant/policy-machinery/blob/d4be292ed9e72de1cd8e04cd57c46415339336b2/controller/controller.go#L273), when the snapshot contains zero updates, then we just `continue`?

Another option is, `if len(events) == 0 { continue }` [here](https://github.com/Kuadrant/policy-machinery/blob/d4be292ed9e72de1cd8e04cd57c46415339336b2/controller/controller.go#L312).